python_modules/dagster/dagster/_core/definitions/asset_subset.py

```diff
@@ -120,12 +124,47 @@ def from_asset_partitions_set(
                 ),
             )
 
+    def valid_subset(self, partitions_def: Optional[PartitionsDefinition]) -> "AssetSubset":
+        """Returns the asset subset representing only the asset partitions which are valid for the
+        current partitions def. Because AssetSubsets may be serialized at a point in time when the
+        set of valid partitions is different from the current set of valid partitions, this check
+        is sometimes necessary.
+        """
+        if self.is_partitioned:
+            if partitions_def is None:
+                return AssetSubset.empty(self.asset_key, partitions_def)
+            # handle the case where a TimeWindowPartitionsDefinition has changed between
+            # serialization time and now
+            elif isinstance(self.value, TimeWindowPartitionsSubset):
+                if (
+                    not isinstance(partitions_def, TimeWindowPartitionsDefinition)
+                    or partitions_def.cron_schedule != self.value.partitions_def.cron_schedule
+                    or any(
+                        time_window.start < partitions_def.start
```

Comment:
IIUC this logic would incorrectly return an empty subset in certain situations when some partitions are still considered valid, i.e. if start date originally at A is changed to a later start date B.

If the subset contains partitions between A and B and after B, technically partitions in B should be returned by this function. But instead we just return an empty subset.

If this is the intended behavior (return an empty subset if the partitions def changes), would prefer a more explicit name to denote this behavior. 

Here we could also compare the serializable unique ID of the partitions subset's time partitions def against the current partitions def's serializable unique ID. This would be preferable to avoid maintaining custom logic that checks partitions defs fields.

Reply:
you're totally right -- I think the ideal here would be something that filtered the subset for only the currently-valid partitions across all possible AssetSubset types, but I realized that this isn't actually necessary at the moment, and I can just go with something like what you mentioned where I just set the subset to None in the case that the partitions def has changed.